File: apps/android-merchant/app/src/main/java/com/tucker/merchant/ui/products/ProductsViewModel.kt
package com.tucker.merchant.ui.products

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.tucker.merchant.data.model.Product
import com.tucker.merchant.data.repository.MerchantRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ProductsViewModel @Inject constructor(
    private val repository: MerchantRepository
) : ViewModel() {

    private val _products = MutableStateFlow<List<Product>>(emptyList())
    val products: StateFlow<List<Product>> = _products.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    init {
        loadProducts()
    }

    fun loadProducts() {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            try {
                _products.value = repository.getProducts()
            } catch (e: Exception) {
                _error.value = e.message ?: "Failed to load products"
            }
            _isLoading.value = false
        }
    }

    fun toggleAvailability(product: Product) {
        viewModelScope.launch {
            try {
                val updated = repository.toggleProductAvailability(product.id, !product.isAvailable)
                _products.value = _products.value.map {
                    if (it.id == product.id) updated else it
                }
            } catch (e: Exception) {
                _error.value = e.message ?: "Failed to update product"
            }
        }
    }

    fun deleteProduct(productId: String) {
        viewModelScope.launch {
            try {
                repository.deleteProduct(productId)
                _products.value = _products.value.filter { it.id != productId }
            } catch (e: Exception) {
                _error.value = e.message ?: "Failed to delete product"
            }
        }
    }
}

File: apps/ios-merchant/TuckerMerchant/Features/Products/ProductsViewModel.swift
import Foundation

@MainActor
class ProductsViewModel: ObservableObject {
    @Published var products: [Product] = []
    @Published var isLoading = false
    @Published var error: String?

    func loadProducts() {
        isLoading = true
        error = nil

        Task {
            do {
                self.products = try await APIService.shared.getMerchantProducts()
            } catch {
                self.error = error.localizedDescription
            }
            isLoading = false
        }
    }

    func toggleAvailability(_ product: Product) {
        Task {
            do {
                let updated = try await APIService.shared.toggleProductAvailability(
                    id: product.id,
                    isAvailable: !product.isAvailable
                )

                if let index = products.firstIndex(where: { $0.id == product.id }) {
                    products[index] = updated
                }
            } catch {
                self.error = error.localizedDescription
            }
        }
    }

    func deleteProduct(_ product: Product) {
        Task {
            do {
                try await APIService.shared.deleteProduct(id: product.id)
                products.removeAll { $0.id == product.id }
            } catch {
                self.error = error.localizedDescription
            }
        }
    }
}
